File: benchmarks/micro/GlobalScalarLoadForwarding.c
#include <stdint.h>
#include <unistd.h>

static int64_t counter = 1;

static int64_t accumulate(int64_t salt) {
    counter += salt;
    int64_t first = counter;
    int64_t second = counter;
    return first + second;
}

int main(void) {
    counter = ((int64_t)getpid() % 31) + 1;
    int64_t checksum = 0;

    for (int32_t i = 0; i < 200000; i += 1) {
        checksum += accumulate((int64_t)(i % 97));
        checksum %= 1000000007;
        counter %= 1000000007;
    }

    return checksum <= 0 ? 1 : 0;
}

File: benchmarks/micro/GlobalScalarLoadForwarding.rs
static mut COUNTER: i64 = 1;

fn accumulate(salt: i64) -> i64 {
    unsafe {
        COUNTER += salt;
        let first = COUNTER;
        let second = COUNTER;
        first + second
    }
}

fn main() {
    unsafe {
        COUNTER = (std::process::id() as i64 % 31) + 1;
    }

    let mut checksum = 0_i64;

    for i in 0_i32..200_000 {
        checksum += accumulate((i % 97) as i64);
        checksum %= 1_000_000_007;
        unsafe {
            COUNTER %= 1_000_000_007;
        }
    }

    if checksum <= 0 {
        std::process::exit(1);
    }
}
